arangod/Pregel/Algos/AIR/VertexComputation.cpp

```diff
@@ -0,0 +1,602 @@
+      getReportManager()
+              .report(ReportLevel::ERR)
+              .with("pregel-id", pregelId())
+              .with("vertex", vertexData()._documentId)
+              .with("phase", phase.name)
+              .with("global-superstep", globalSuperstep())
+              .with("phase-step", phaseGlobalSuperstep())
+              .with("message", msg->_value.toJson())
+              .with("sender", msg->_sender)
+              .with("accumulator", accumName)
+          << "in phase `" << phase.name << "` processing incoming messages for accumulator `"
+          << accumName << "` failed: " << res.error().toString();
+      return std::move(res.error());
+    }
+
+    accumChanged |= res.value() == AccumulatorBase::UpdateResult::CHANGED;
+  }
+  return accumChanged;
+}
+
+greenspun::EvalResult VertexComputation::runProgram(greenspun::Machine& ctx, VPackSlice program) {
+  VPackBuilder resultBuilder;
+
+  // A valid pregel program can at the moment return one of five values: none,
+  // true, false, "vote-halt", or "vote-active"
+  //
+  // if it returns none, false, or "vote-halt", then we voteHalt(), if it
+  // returns true or "vote-active" we voteActive()
+  //
+  // In all other cases we throw an error
+  auto evaluateResult = [this](VPackSlice& rs) -> greenspun::EvalResult {
+    if (rs.isNone()) {
+      this->voteHalt();
+      return {};
+    } else if (rs.isBoolean()) {
+      if (rs.getBoolean()) {
+        this->voteActive();
+        return {};
+      } else {
+        this->voteHalt();
+        return {};
+      }
+    } else if (rs.isString()) {
+      if (rs.stringRef().equals("vote-active")) {
+        this->voteActive();
+        return {};
+      } else if (rs.stringRef().equals("vote-halt")) {
+        this->voteHalt();
+        return {};
+      }
+    }
+    // Not a valid value, vote to halt and return error
+    voteHalt();
+    return greenspun::EvalError("pregel program returned " + rs.toJson() +
+                                ", expect one of `none`, `true`, `false`, "
+                                "`\"vote-halt\", or `\"vote-active\"`");
+  };
+
+  auto evalResult = Evaluate(ctx, program, resultBuilder);
+  if (!evalResult) {
+    // An error occurred during execution, we vote halt and return the error
+    voteHalt();
+    return evalResult.mapError(
+        [](greenspun::EvalError& err) { err.wrapMessage("at top-level"); });
+  } else {
+    auto rs = resultBuilder.slice();
+    return evaluateResult(rs);
+  }
+}
+
+void VertexComputation::compute(MessageIterator<MessageData> const& incomingMessages) {
+  auto phase_index = *getAggregatedValue<uint32_t>("phase");
+  auto phase = _algorithm.options().phases.at(phase_index);
+
+  /*LOG_DEVEL << "running phase " << phase.name
+            << " superstep = " << phaseGlobalSuperstep()
+            << " global superstep = " << globalSuperstep() << " at vertex "
+            << vertexData()._vertexId;*/
+
+  std::size_t phaseStep = phaseGlobalSuperstep();
```

Comment:
```suggestion
  auto phaseStep = phaseGlobalSuperstep();
```
`phaseGlobalSuperstep` does not return a `size_t`.